experimental/InjectiveResolutions/src/MonoidAlgebra.jl

```diff
@@ -0,0 +1,611 @@
+  prime::Union{MPolyIdeal,MPolyQuoIdeal} #homogeneous prime corresponding to face
+  poly::Polyhedron  #face as polyhedron
+end
+
+struct HyperplaneQ # a hyperplane bounding the cone RR_{\geq 0}Q
+  hyperplane::Polyhedron
+  A::Matrix{Int}
+  b::Vector{Int}
+end
+@attributes mutable struct MonoidAlgebra{CoeffType,AlgebraType} <: Ring # monoid algebra with associated data
+  algebra::AlgebraType
+  pointed::Union{Nothing,Bool} #cone pointed
+  polyhedral_cone::Cone{QQFieldElem}
+  cone::Polyhedron
+  faces::Vector{FaceQ}
+  hyperplanes::Vector{HyperplaneQ}
+  zonotope::Tuple{Polyhedron,Vector{Int}}
+
+  function MonoidAlgebra(
+    A::AlgebraType; check::Bool=true
+  ) where {AlgebraType<:Union{MPolyRing,MPolyQuoRing}}
+    #check if monoid algebra
+    @check is_zm_graded(A) "given algebra is not ZZ^d-graded"
+    gg_Q = grading_group(A)
+    @check is_free(gg_Q) && is_abelian(gg_Q) "given algebra is not a monoid algebra"
+    kk = coefficient_ring(A)
+    result = new{elem_type(kk),AlgebraType}(A, nothing)
+  end
+end
+
+is_graded(A::MonoidAlgebra) = true
+is_zm_graded(A::MonoidAlgebra) = true
+
+function polyhedral_cone(A::MonoidAlgebra)
+  if !isdefined(A, :polyhedral_cone)
+    A.polyhedral_cone = get_polyhedral_cone(A.algebra)
+  end
+  return A.polyhedral_cone
+end
+
+@doc raw"""
+    cone(A::MonoidAlgebra)
+
+Given a monoid algebra with underlying monoid $Q$, this function returns the polyhedral cone $\mathbb{R}_{\geq 0}Q$ as a polyhedron. 
+
+"""
+function cone(A::MonoidAlgebra)
+  if !isdefined(A, :cone)
+    A.cone = polyhedron(polyhedral_cone(A))
+  end
+  return A.cone
+end
+
+
+@doc raw"""
+    faces(A::MonoidAlgebra)
+
+Given a monoid algebra with underlying monoid $Q$, this function a list of all faces of the polyhedral cone $\mathbb{R}_{\geq 0}Q$ with their corresponding homogeneous prime ideals. 
+"""
+function faces(A::MonoidAlgebra)
+  if !isdefined(A, :faces)
+    A.faces = get_faces_of_polyhedral_cone(A.algebra, zonotope(A)[1], cone(A))
+  end
+  return A.faces
+end
+
+function hyperplanes(A::MonoidAlgebra)
+  if !isdefined(A, :hyperplanes)
+    A.hyperplanes = get_bounding_hyperplanes(cone(A))
+  end
+  return A.hyperplanes
+end
+
+function is_pointed(A::MonoidAlgebra)
+  if isnothing(A.pointed)
+    A.pointed = is_pointed(polyhedral_cone(A))
+  end
+  return A.pointed::Bool
+end
+
+function zonotope(A::MonoidAlgebra)
+  if !isdefined(A, :zonotope)
+    A.zonotope = get_zonotope(cone(A))
+  end
+  return A.zonotope
+end
+
+coefficient_ring(A::MonoidAlgebra) = coefficient_ring(A.algebra)
+number_of_variables(A::MonoidAlgebra) = ngens(A.algebra)
+
+### Elements of MonoidAlgebras
+mutable struct MonoidAlgebraElem{CoeffType,ParentType} <: RingElem
+  parent::ParentType
+  elem::RingElem
+  # TODO: Do we want to store additional information on the elements here?
+
+  function MonoidAlgebraElem(
+    A::ParentType
+  ) where {CoeffType,ParentType<:MonoidAlgebra{CoeffType}}
+    return new{CoeffType,ParentType}(A)
```

Comment:
Where do you need this? Couldn't it be
```suggestion
    return new{CoeffType,ParentType}(A, zero(A.algebra))
```
Then you could remove a ton of `isdefined` special cases below?